polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp

```diff
@@ -478,33 +478,33 @@ static bool canBeHoisted(LICMCandidate &candidate, LoopLikeOpInterface loop,
   if (isMemoryEffectFree(&op)) {
     LLVM_DEBUG({
       llvm::dbgs() << "Operation: " << op << "\n";
-      llvm::dbgs().indent(2) << "**** can be hoisted: has no side effects\n\n";
+      llvm::dbgs().indent(2) << "**** has no side effects\n\n";
     });
-    return true;
-  }
+  } else {
 
-  // Do not hoist operations that allocate a resource.
-  const OperationSideEffects sideEffects(op, aliasAnalysis, domInfo);
-  if (sideEffects.allocatesResource()) {
-    LLVM_DEBUG({
-      llvm::dbgs() << "Operation: " << op << "\n";
-      llvm::dbgs().indent(2)
-          << "**** cannot be hoisted: operation allocates a resource\n\n";
-    });
-    return false;
-  }
+    // Do not hoist operations that allocate a resource.
+    const OperationSideEffects sideEffects(op, aliasAnalysis, domInfo);
+    if (sideEffects.allocatesResource()) {
+      LLVM_DEBUG({
+        llvm::dbgs() << "Operation: " << op << "\n";
+        llvm::dbgs().indent(2)
+            << "**** cannot be hoisted: operation allocates a resource\n\n";
+      });
+      return false;
+    }
 
-  LLVM_DEBUG(llvm::dbgs() << sideEffects);
+    LLVM_DEBUG(llvm::dbgs() << sideEffects);
 
-  // If the operation has side effects, check whether other operations in the
-  // loop prevent hosting it.
-  if ((sideEffects.readsFromResource() || sideEffects.writesToResource() ||
-       sideEffects.freesResource()) &&
-      hasConflictsInLoop(candidate, loop, willBeMoved, aliasAnalysis,
-                         domInfo)) {
-    LLVM_DEBUG(llvm::dbgs().indent(2)
-               << "**** cannot be hoisted: found conflicting operation\n\n");
-    return false;
+    // If the operation has side effects, check whether other operations in the
+    // loop prevent hosting it.
+    if ((sideEffects.readsFromResource() || sideEffects.writesToResource() ||
+         sideEffects.freesResource()) &&
+        hasConflictsInLoop(candidate, loop, willBeMoved, aliasAnalysis,
+                           domInfo)) {
+      LLVM_DEBUG(llvm::dbgs().indent(2)
+                 << "**** cannot be hoisted: found conflicting operation\n\n");
+      return false;
+    }
   }
 
   // Recurse into the regions for this op and check whether the contained ops
```

Comment:
Can we avoid recursing into the regions, if the operation is `IsolatedFromAbove`?

Reply:
Maybe @etiotto and @whitneywhtsang can throw some light here, but I think the loop below is needed in case the operation may have recursive memory effects in order to check this. Am I right?

Reply:
> IsolatedFromAbove trait signals that the regions of an operations are known to be isolated
from above. This trait asserts that the regions of an operation will not
capture, or reference, SSA values defined above the region scope.

So we can keep the `isMemoryEffectFree` early exit, but we also need to check that it has `IsolatedFromAbove` trait when it is a region operation.

Reply:
I agree with @whitneywhtsang and @sommerlukas.

Reply:
I added an early exit if the operation has the `IsolatedFromAbove` trait. Testing the early exit would involve importing an additional dialect, as no operations in the dialects we're currently handling have the `IsolatedFromAbove` and `HasRecursiveMemoryEffects` traits.